# Pair Programming ANOVA

    - En el pair programming de hoy usaremos el set de datos que guardastéis en el pair programming de normalización y estandarización.
    - Hasta ahora habéis estado evaluando las características de vuestro set de datos y habéis hecho una gran exploración, es el momento de hacer vuestro primer ANOVA! En el ejercicio de hoy tendréis que hacer un ANOVA con vuestro datos y hacer una interpretación de los resultados.
    - 📌 NOTA Puede que vuestros datos no se ajusten o no cumplan todas las asunciones, no pasa nada, haced el ANOVA e interpretad los resultados. En próximas lecciones aprenderemos que podemos hacer cuando nos encontramos en esta situación.

In [1]:
# Tratamiento de datos
# -----------------------------------------------------------------------
import numpy as np
import pandas as pd
import random 

# Gráficos
# ------------------------------------------------------------------------------
import matplotlib.pyplot as plt
import seaborn as sns

# Estadísticos
# ------------------------------------------------------------------------------
import statsmodels.api as sm
from statsmodels.formula.api import ols
from statsmodels.multivariate.manova import MANOVA
from sklearn.preprocessing import StandardScaler

plt.rcParams["figure.figsize"] = (10,8) 

/home/cristina/anaconda3/lib/python3.9/site-packages/statsmodels/tsa/base/tsa_model.py:7: FutureWarning: pandas.Int64Index is deprecated and will be removed from pandas in a future version. Use pandas.Index with the appropriate dtype instead.
  from pandas import (to_datetime, Int64Index, DatetimeIndex, Period,
/home/cristina/anaconda3/lib/python3.9/site-packages/statsmodels/tsa/base/tsa_model.py:7: FutureWarning: pandas.Float64Index is deprecated and will be removed from pandas in a future version. Use pandas.Index with the appropriate dtype instead.
  from pandas import (to_datetime, Int64Index, DatetimeIndex, Period,


In [2]:
df = pd.read_csv('../datos/supermarket_robust.csv', index_col = 0)
df.head(1)

Invoice ID Branch    City Customer type  Gender       Product line  \
0  750-67-8428      A  Yangon        Member  Female  Health and beauty   

   Unit price  Quantity    Tax 5%     Total  Payment  Rating hora_compra  \
0    0.431869       0.4  0.850677  0.850677  Ewallet     9.1      Mañana   

  mes_compra  RATING_BOX  RATING_LOG  
0    January    5.620596    2.208274

In [3]:
columnas_nuevas = {col: col.replace(' ', '_').replace('%', '') for col in df.columns}

In [4]:
df.rename(columns = columnas_nuevas, inplace = True)

In [5]:
df.head(1)

Invoice_ID Branch    City Customer_type  Gender       Product_line  \
0  750-67-8428      A  Yangon        Member  Female  Health and beauty   

   Unit_price  Quantity     Tax_5     Total  Payment  Rating hora_compra  \
0    0.431869       0.4  0.850677  0.850677  Ewallet     9.1      Mañana   

  mes_compra  RATING_BOX  RATING_LOG  
0    January    5.620596    2.208274

In [6]:
lm = ols('Rating ~ Quantity  + City + Tax_5 +  Total + Unit_price + Customer_type + Gender + Payment + hora_compra + mes_compra' , data=df).fit()
sm.stats.anova_lm(lm)

df       sum_sq   mean_sq         F    PR(>F)
City             2.0    12.233579  6.116789  2.072738  0.126389
Customer_type    1.0     1.118048  1.118048  0.378862  0.538355
Gender           1.0     0.100740  0.100740  0.034137  0.853454
Payment          2.0     0.761346  0.380673  0.128995  0.878993
hora_compra      2.0     4.862230  2.431115  0.823809  0.439059
mes_compra       2.0    10.214200  5.107100  1.730594  0.177717
Quantity         1.0     0.706522  0.706522  0.239412  0.624740
Tax_5            1.0     3.826289  3.826289  1.296578  0.255116
Total            1.0     0.021913  0.021913  0.007425  0.931348
Unit_price       1.0     7.079440  7.079440  2.398942  0.121738
Residual       986.0  2909.752812  2.951068       NaN       NaN

In [7]:
lm2 = ols('RATING_BOX ~ Quantity  + City + Tax_5 +  Total + Unit_price + Customer_type + Gender + Payment + hora_compra + mes_compra' , data=df).fit()
sm.stats.anova_lm(lm2)

df       sum_sq   mean_sq         F    PR(>F)
City             2.0     4.626220  2.313110  2.076485  0.125918
Customer_type    1.0     0.476982  0.476982  0.428188  0.513032
Gender           1.0     0.061306  0.061306  0.055035  0.814572
Payment          2.0     0.272371  0.136186  0.122254  0.884937
hora_compra      2.0     1.928119  0.964059  0.865439  0.421186
mes_compra       2.0     3.796864  1.898432  1.704228  0.182448
Quantity         1.0     0.275673  0.275673  0.247472  0.618972
Tax_5            1.0     1.568449  1.568449  1.408001  0.235674
Total            1.0     0.001107  0.001107  0.000994  0.974859
Unit_price       1.0     2.759318  2.759318  2.477047  0.115840
Residual       986.0  1098.358981  1.113954       NaN       NaN

In [8]:
lm2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:             RATING_BOX   R-squared:                       0.014
Model:                            OLS   Adj. R-squared:                  0.001
Method:                 Least Squares   F-statistic:                     1.087
Date:                Thu, 18 May 2023   Prob (F-statistic):              0.366
Time:                        19:37:14   Log-Likelihood:                -1465.8
No. Observations:                1000   AIC:                             2960.
Df Residuals:                     986   BIC:                             3028.
Df Model:                          13                                         
Covariance Type:            nonrobust                                         
===========================================================================================
                              coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------------
Intercept                   4.3322      0.107     40.385      0.000       4.122       4.543
City[T.Naypyitaw]           0.1672      0.083      2.025      0.043       0.005       0.329
City[T.Yangon]              0.1325      0.082      1.618      0.106      -0.028       0.293
Customer_type[T.Normal]     0.0476      0.067      0.709      0.478      -0.084       0.179
Gender[T.Male]              0.0270      0.068      0.399      0.690      -0.106       0.160
Payment[T.Credit card]      0.0405      0.083      0.488      0.625      -0.122       0.203
Payment[T.Ewallet]         -0.0050      0.081     -0.063      0.950      -0.164       0.153
hora_compra[T.Noche]       -0.0605      0.084     -0.723      0.470      -0.225       0.104
hora_compra[T.Tarde]       -0.0991      0.079     -1.248      0.212      -0.255       0.057
mes_compra[T.January]      -0.0434      0.083     -0.523      0.601      -0.206       0.119
mes_compra[T.March]        -0.1433      0.084     -1.713      0.087      -0.307       0.021
Quantity                    0.2059      0.134      1.541      0.124      -0.056       0.468
Tax_5                      -0.1384      0.071     -1.940      0.053      -0.278       0.002
Total                      -0.1384      0.071     -1.940      0.053      -0.278       0.002
Unit_price                  0.1912      0.122      1.568      0.117      -0.048       0.431
==============================================================================
Omnibus:                      426.476   Durbin-Watson:                   1.984
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               53.122
Skew:                          -0.062   Prob(JB):                     2.92e-12
Kurtosis:                       1.878   Cond. No.                     1.05e+16
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 2.38e-29. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

In [9]:
lm.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 Rating   R-squared:                       0.014
Model:                            OLS   Adj. R-squared:                  0.001
Method:                 Least Squares   F-statistic:                     1.064
Date:                Thu, 18 May 2023   Prob (F-statistic):              0.387
Time:                        19:37:14   Log-Likelihood:                -1953.0
No. Observations:                1000   AIC:                             3934.
Df Residuals:                     986   BIC:                             4003.
Df Model:                          13                                         
Covariance Type:            nonrobust                                         
===========================================================================================
                              coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------------
Intercept                   6.9696      0.175     39.918      0.000       6.627       7.312
City[T.Naypyitaw]           0.2708      0.134      2.015      0.044       0.007       0.535
City[T.Yangon]              0.2170      0.133      1.628      0.104      -0.045       0.479
Customer_type[T.Normal]     0.0737      0.109      0.675      0.500      -0.141       0.288
Gender[T.Male]              0.0382      0.110      0.347      0.729      -0.178       0.254
Payment[T.Credit card]      0.0693      0.135      0.513      0.608      -0.196       0.335
Payment[T.Ewallet]         -0.0058      0.131     -0.044      0.965      -0.264       0.252
hora_compra[T.Noche]       -0.0932      0.136     -0.685      0.494      -0.360       0.174
hora_compra[T.Tarde]       -0.1578      0.129     -1.221      0.223      -0.411       0.096
mes_compra[T.January]      -0.0672      0.135     -0.497      0.619      -0.332       0.198
mes_compra[T.March]        -0.2341      0.136     -1.719      0.086      -0.501       0.033
Quantity                    0.3264      0.217      1.501      0.134      -0.100       0.753
Tax_5                      -0.2197      0.116     -1.892      0.059      -0.448       0.008
Total                      -0.2197      0.116     -1.892      0.059      -0.448       0.008
Unit_price                  0.3055      0.199      1.539      0.124      -0.084       0.695
==============================================================================
Omnibus:                      463.824   Durbin-Watson:                   1.983
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               53.806
Skew:                           0.019   Prob(JB):                     2.07e-12
Kurtosis:                       1.864   Cond. No.                     1.05e+16
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 2.38e-29. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

Tras chequear los resultados del ANOVA, comprobamos que la columna de ciudad que no habíamos considerado hasta ahora, SI puede tener relación en el Rating del cliente. Las variables numéricas como ya habíamos visto no tienen ningún efecto, y las categóricas al contrario de lo que pensábamos tampoco tienen ningun efecto.

Vemos que tenemos un R-squared de 0.014: el resultado es muy bajo, por lo que concluimos que las variables predictoras no explican nuestra variable respuesta ('Rating').